### PR TITLE
blockdev_mirror:add nbd support for block mirror

### PR DESCRIPTION
**Pull request by aliang123**
Add backend:nbd support for block mirror

Signed-off-by: Aihua Liang <aliang@redhat.com>
id:2068982

### CODE REVIEW

**Comment by aliang123** [comment]
(01/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_multiple_blocks.q35: PASS (274.31 s)
(02/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_same_src_tgt.q35: PASS (29.79 s)
(03/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.with_data_plane.q35: PASS (93.67 s)
(04/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.q35: PASS (98.18 s)
(05/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_512.with_data_plane.q35: PASS (165.32 s)
(06/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_512.q35: PASS (171.01 s)
(07/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_2M.with_data_plane.q35: PASS (115.53 s)
(08/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_2M.q35: PASS (112.83 s)
(09/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_random_cluster_size.with_data_plane.q35: PASS (102.19 s)
(10/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_random_cluster_size.q35: PASS (100.50 s)
 (11/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_with_ignore.q35: PASS (116.70 s)
(12/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_stress.q35: PASS (215.57 s)
(13/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_vm_reboot.q35: PASS (220.81 s)
(14/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_forbidden_actions.q35: PASS (62.81 s)
(15/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_sync_none.q35: PASS (183.19 s)
(16/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_readonly.q35: PASS (54.42 s)
 (17/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_filternode.q35: PASS (100.86 s)
(18/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_firewall.q35: PASS (249.92 s)
(19/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_error.q35: PASS (56.29 s)
 (20/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_hotunplug.q35: PASS (144.72 s)
(21/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_ready_vm_down.q35: PASS (59.30 s)
(22/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_cancel_ready_job_with_io.q35: PASS (66.06 s)
(23/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_vm_stop_cont.q35: PASS (370.46 s)
(24/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_complete_running_job.q35: PASS (63.78 s)
(25/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.granularity_512.q35: PASS (105.93 s)
(26/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.granularity_64M.q35: PASS (131.15 s)
(27/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.granularity_between_512_64M.buf_size_default.q35: PASS (101.89 s)
(28/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.granularity_between_512_64M.buf_size_random.q35: PASS (111.11 s)
(29/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.post_job_running.auto_dismiss_on.auto_finalize_on.q35: PASS (101.27 s)
(30/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.post_job_running.auto_dismiss_on.auto_finalize_off.q35: PASS (104.71 s)
(31/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.post_job_running.auto_dismiss_off.auto_finalize_on.q35: PASS (101.81 s)
(32/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_simple.post_job_running.auto_dismiss_off.auto_finalize_off.q35: PASS (102.49 s)
(33/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_install.q35: PASS (758.77 s)
(34/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_sync_top.q35: PASS (223.79 s)
(35/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_cancel_running_job.q35: PASS (54.15 s)
 (36/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_remote_server_down.q35: PASS (148.66 s)
(37/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space.q35: PASS (47.57 s)
(38/38) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_speed.q35: PASS (78.77 s)

**Comment by aliang123** [comment]
@zhencliu @ngu-niny or @MiriamDeng, please help to review, thanks.

File: qemu/tests/cfg/blockdev_mirror_cancel_ready_job_with_io.cfg [hunk 1]
```diff
@@ -30,7 +30,8 @@
     image_name_mirror1 = mirror1
 
     nbd:
-        nbd_port_data1 = 10822
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_mirror_cancel_ready_job_with_io.cfg:33]: 10822 is used for a 2g qcow2 nbd image, so here do we want to add that? e.g.
qcow2:
    image_format_data1 = qcow2
    nbd_port_data1 = 10822
raw:
   nbd_port_data1 = 10831
   image_format_data1 = raw

Comment (zhencliu) [reply qemu/tests/cfg/blockdev_mirror_cancel_ready_job_with_io.cfg:33]: I noticed that we have several same changes(10822->10831) for the following cfg files, please also consider those changes if you agree to add qcow2 support, I will not leave comments for those changes, thanks.

File: qemu/tests/cfg/blockdev_mirror_install.cfg [hunk 2]
```diff
@@ -63,7 +63,7 @@
         qcow2:
             nbd_port_image1 = 10820
         raw:
-            nbd_port_image1 = 10819
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_mirror_install.cfg:66]: 10819 is also used for a raw nbd image:
local image format: raw
export format: raw
VS.
10830:
local image format: qcow2
export format: qcow2

Not sure which is popular use in product, here clarify it, just for your reference.

**Comment by MiriamDeng** [comment]
Running it and will update the results later.

**Comment by MiriamDeng** [comment]
Most test cases passed on ppc
...
 (30/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_2M.with_data_plane: STARTED
| (30/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_2M.with_data_plane: PASS (167.00 s)
 (31/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_2M: STARTED
- (31/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_2M: PASS (157.25 s)
 (32/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_random_cluster_size.with_data_plane: STARTED
\ (32/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_random_cluster_size.with_data_plane: PASS (413.98 s)
 (33/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_random_cluster_size: STARTED
- (33/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_random_cluster_size: PASS (146.51 s)
 (34/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_stress: STARTED
| (34/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_stress: PASS (366.99 s)
 (35/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_readonly: STARTED
| (35/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_readonly: PASS (83.35 s)
 (36/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_sync_none: STARTED
\ (36/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_sync_none: PASS (294.91 s)
 (37/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_ready_vm_down: STARTED
- (37/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_ready_vm_down: PASS (102.64 s)
 (38/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space: STARTED
- (38/38) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space: PASS (63.52 s)
There are three failures:
Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_install) FAIL
Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_512.with_data_plane) FAIL
Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_full_mirror.simple_test.dst_cluster_size_512) FAIL AssertionError('wait for block job complete event timeout in 600 seconds',)

**Comment by zhencliu** [comment]
LGTM

**Pull request merged**